### PR TITLE
[Feature] Add ESPNetV1

### PR DESCRIPTION
**Pull request by simuler**
添加ESPNet模型（Dynamic Multi-scale Filters for Semantic Segmentation）

添加内容：
configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml
configs/espnetv1/README.md
paddleseg/models/espnetv1.py
paddleseg/models/init.py

### CODE REVIEW

File: configs/espnetv1/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# ESPNet: Efficient Spatial Pyramid of Dilated Convolutions for Semantic Segmentation
+
+
+## Reference
+[1] Mehta S ,  Rastegari M ,  Caspi A , et al. ESPNet: Efficient Spatial Pyramid of Dilated Convolutions for Semantic Segmentation[J]. Springer, Cham, 2018.
```
Comment (shiyutang) [inline configs/espnetv1/README.md]: Plz change this quote to Chicago format, which includes their full name.
you can refer to this: https://github.com/PaddlePaddle/PaddleSeg/blob/develop/docs/pr/pr/style_cn.md#%E4%B8%89%E6%96%B0%E5%A2%9E-pr-checklist

File: configs/espnetv1/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+## Performance
+CityScapes
```
Comment (shiyutang) [inline configs/espnetv1/README.md]: ### Cityscapes

File: configs/espnetv1/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+| Model|steps|opt|image_size|batch_size|dataset|memory|card|mIou|config|
+| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
+|ESPNet|120k|adam|1024x512|4|CityScapes|32G|4|0.6365|[espnet_cityscapes_1024x512_120k.yml](configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml)|
```
Comment (shiyutang) [inline configs/espnetv1/README.md]: Actually, you do not need to include so much information. Please refer to our style MD before you bring up your PR

File: paddleseg/models/espnetv1.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+from paddleseg.utils import utils
+
+def SyncBatchNorm(*args, **kwargs):
+    """In cpu environment nn.SyncBatchNorm does not have kernel so use nn.BatchNorm2D instead"""
+    if paddle.get_device() == 'cpu' or os.environ.get('PADDLESEG_EXPORT_STAGE'):
+        return nn.BatchNorm2D(*args, **kwargs)
+    elif paddle.distributed.ParallelEnv().nranks == 1:
+        return nn.BatchNorm2D(*args, **kwargs)
+    else:
+        return nn.SyncBatchNorm(*args, **kwargs)
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: please make use of the same function we have in layers.

File: paddleseg/models/espnetv1.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        self.out_up = nn.Conv2DTranspose(num_classes, num_classes, 2, stride=2, padding=0, output_padding=0, bias_attr=False)
+    
+    def forward(self, x):
+        p1, p2, p3 = self.encoder(x) # shape [N, C, H/2, W/2]  [N, C, H/4, W/4]  [N, C, H/8, W/8] 
+        up_p3 = self.level3_up(p3) # [N, C, H/4, W/4]
+
+        combine = self.combine_l2_l3(paddle.concat([up_p3, p2], axis=1))
+        up_p2 = self.level2_up(combine)  # [N, C, H/2, W/2]
+
+        combine = self.out_proj(paddle.concat([up_p2, p1], axis=1))  # shape [N, C, H/2, W/2]
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: Remove this annotation please, here and elsewhere

File: paddleseg/models/espnetv1.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+ 
+@manager.MODELS.add_component
+class ESPNetV1(nn.Layer):
+    """
+    The ESPNetV1 implementation based on PaddlePaddle.
+    The original article refers to
+    Sachin Mehta1, Mohammad Rastegari, Anat Caspi, Linda Shapiro, and Hannaneh Hajishirzi. "ESPNet: Efficient Spatial Pyramid of Dilated Convolutions for Semantic Segmentation"
+    (https://arxiv.org/abs/1803.06815).
+    Args:
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py:33]: Add blank lines plz

**Review by shiyutang — CHANGES_REQUESTED** [review]
Please read our style markdown before PR. This PR needs more adjustment than my comments.
Style markdown link: https://github.com/PaddlePaddle/PaddleSeg/blob/develop/docs/pr/pr/style_cn.md#paddleseg%E6%A8%A1%E5%9E%8B%E5%BC%80%E5%8F%91%E8%A7%84%E8%8C%83

File: configs/espnetv1/README.md (hunk 1 above)
Comment (simuler) [reply configs/espnetv1/README.md]: fixed

File: configs/espnetv1/README.md (hunk 2 above)
Comment (simuler) [reply configs/espnetv1/README.md]: fixed

File: configs/espnetv1/README.md (hunk 3 above)
Comment (simuler) [reply configs/espnetv1/README.md]: fixed

File: paddleseg/models/espnetv1.py (hunk 4 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: fixed

File: paddleseg/models/espnetv1.py (hunk 5 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: fixed

File: paddleseg/models/espnetv1.py (hunk 6 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py:33]: fixed

File: configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+_base_: '../_base_/cityscapes.yml'
+
+batch_size: 4
+iters: 120000
+
+
```
Comment (shiyutang) [inline configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml]: remove this plz

File: paddleseg/models/espnetv1.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        feat3 = paddle.concat([f3, f3_res], axis=1)
+        feat3 = self.br3(feat3)
+        p3 = self.proj3(feat3)
+
+        return p1, p2, p3
+
+
+if __name__ == '__main__':
+    model = ESPNetV1(19, 3, 2, 8)
+    paddle.summary(model, (4, 3, 256, 256))
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: remove test code plz

File: paddleseg/models/espnetv1.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+class ConvDilated(nn.Layer):
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1, dilation=1):
+        super().__init__()
+        padding = int((kernel_size - 1) / 2) * dilation
+        self.conv = nn.Conv2D(in_channels, out_channels, kernel_size, stride=stride, padding=padding, dilation=dilation, bias_attr=False)
+
+    def forward(self, x):
+        x = self.conv(x)
+        return x
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: It seems these two function is not that necessary

File: paddleseg/models/espnetv1.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+import os
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: This should be moved up

File: paddleseg/models/espnetv1.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1):
+        super().__init__()
+        padding = int((kernel_size - 1) / 2)
+        self.conv = nn.Conv2D(in_channels, out_channels, kernel_size, stride=stride, padding=padding, bias_attr=False)
+        self.bn = layers.SyncBatchNorm(out_channels)
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        return x
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: please checkout layers package for these function

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: paddleseg/models/espnetv1.py (hunk 11 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: 1、The ConvBNPReLU of layers package cannot modified the stride
`class ConvBNPReLU(nn.Layer):
    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size,
                 padding='same',
                 **kwargs):
        super().__init__()

        self._conv = nn.Conv2D(in_channels,
                               out_channels,
                               kernel_size,
                               padding=padding,
                               **kwargs)

        if 'data_format' in kwargs:
            data_format = kwargs['data_format']
        else:
            data_format = 'NCHW'
        self._batch_norm = SyncBatchNorm(out_channels, data_format=data_format)
        self._prelu = layers.Activation("prelu")

    def forward(self, x):
        x = self._conv(x)
        x = self._batch_norm(x)
        x = self._prelu(x)
        return x`

2、no BNPReLU  in   layers package

3、 I checked the code again, ConvBN is not used

File: configs/espnetv1/README.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+## Reference
+
+> Mehta Sachin, Mohammad Rastegari, Anat Caspi, Linda Shapiro, and Hannaneh Hajishirzi. "ESPNet: Efficient Spatial Pyramid of Dilated Convolutions for Semantic Segmentation.".In Proceedings of the European Conference on Computer Vision, pp. 552-568. 2018.
+
+## Performance
+
+### Cityscapes
+
+| Model | Backbone | Resolution | Training Iters | mIoU | Links |
```
Comment (shiyutang) [inline configs/espnetv1/README.md]: please add the result of mIOU(+flip) and +ms & flip like this:
image
You can get the result through:
```bash 
flip+ms
python3  val.py --config .yml --model_path best_model/model.pdparams --num_workers 2 --aug_eval --flip_horizontal --scales 0.75 1.0 1.25

flip
python3  val.py --config config.yml --model_path best_model/model.pdparams --num_workers 2 --aug_eval --flip_horizontal 
```

File: paddleseg/models/espnetv1.py (hunk 10 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: fixd

File: paddleseg/models/espnetv1.py (hunk 9 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: fixd

File: paddleseg/models/espnetv1.py (hunk 8 above)
Comment (simuler) [reply paddleseg/models/espnetv1.py]: fixd

File: configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml (hunk 7 above)
Comment (simuler) [reply configs/espnetv1/espnetv1_cityscapes_1024x512_120k.yml]: fixd

File: paddleseg/models/espnetv1.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        padding = int((kernel_size - 1) / 2)
+        self.conv = nn.Conv2D(in_channels, out_channels, kernel_size, stride=stride, padding=padding, bias_attr=False)
+        self.bn = layers.SyncBatchNorm(out_channels)
+        self.act = nn.PReLU(out_channels)
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        x = self.act(x)
+        return x
```
Comment (shiyutang) [inline paddleseg/models/espnetv1.py]: Actually the models.layers.ConvBNPReLU can pass in kwargs to change the padding and stride of Conv2D.

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: paddleseg/models/espnetv1.py (hunk 13 above)
Comment (justld) [reply paddleseg/models/espnetv1.py]: fixed

File: configs/espnetv1/README.md (hunk 12 above)
Comment (justld) [reply configs/espnetv1/README.md]: README.md is updated.  Will you train espnetv1 from the scratch? If you do, wo don't supply weight trained before.

Comment (shiyutang) [reply configs/espnetv1/README.md]: we will not train it again, you can give the model's weight to me.

Comment (justld) [reply configs/espnetv1/README.md]: Done

**Review by shiyutang — APPROVED** [review]

**Review by michaelowenliu — APPROVED** [review]
